### PR TITLE
Add tests for `sycl_ext_oneapi_composite_device` extension

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Test plan: KhronosGroup/SYCL-CTS#865

### CODE REVIEW

File: tests/extension/oneapi_composite_device/aspects.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/aspects.cpp:45]: Should we also check that the devices returned by `platform::get_composite_devices` are present in the result of the free function `get_composite_devices`?

File: tests/extension/oneapi_composite_device/descendent_device.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/descendent_device.cpp:36]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/aspects.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/aspects.cpp:84]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/aspects.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/aspects.cpp:38]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/device_info_descriptors.cpp:37]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/device_info_descriptors.cpp:90]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/device_info_descriptors.cpp:123]:
```suggestion
            "device associated with it");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:38]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:60]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:65]:
```suggestion
  // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:76]:
```suggestion
    // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:95]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:101]:
```suggestion
  // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:112]:
```suggestion
    // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:132]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: Why do we need this? I thought `sycl::device::get_devices()` should already return the devices returned by `platform.get_devices()`, am I wrong?

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:146]: Similar to the above one: the free function should include all the composite devices returned by `platform.ext_oneapi_get_composite_devices`, why do we need them both?

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/more_complex_test_cases.cpp:36]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/more_complex_test_cases.cpp:86]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/descendent_device.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/descendent_device.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/aspects.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/aspects.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/aspects.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/aspects.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/more_complex_test_cases.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/more_complex_test_cases.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (maarquitos14) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: I don't understand what you mean, sorry. The highlighted piece of code firstly gets all the devices of the platform, and subsequently it adds all the devices returned by `sycl::device::get_devices()` to the same vector. What I mean is that I think there will be duplicates here. If we do add all the devices from `sycl::device::get_devices()`, why do we need to start with `platform.get_devices()`? Like I said, I would expect duplicates here.

File: tests/extension/oneapi_composite_device/CMakeLists.txt
Comment (bader) [inline tests/extension/oneapi_composite_device/CMakeLists.txt:7]:
```suggestion
    aspects.cpp
    descendent_device.cpp
    device_info_descriptors.cpp
    enumerating_composite_devices.cpp
    more_complex_test_cases.cpp
```

Comment (bader) [reply tests/extension/oneapi_composite_device/CMakeLists.txt]: Done.

**Comment by maarquitos14** [comment]
@bader Other than the current unresolved threads, this LGTM, but I'd like to have either some discussion or resolution for the unresolved threads before approving.

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: Apology, I misread the question and mistakenly though that it was asked about a different test.

I agree that `platform.get_devices()` should contain everything that `device::get_devices()` returns, so we can probably simplify the code here

Comment (maarquitos14) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: > I agree that platform.get_devices() should contain everything that device::get_devices() returns, so we can probably simplify the code here

I think it's actually the other way around, isn't it? The result of `device::get_devices()` should contain the result of `platform.get_devices()`.

Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: > I think it's actually the other way around, isn't it? The result of `device::get_devices()` should contain the result of `platform.get_devices()`.

Right :)

File: tests/extension/oneapi_composite_device/aspects.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/aspects.cpp:45]: Added this test case into "enumerating composite devices" in bcebe1ad87778bca975673cc67eafaf1a1c23c26

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: Applied in f8b1ca288cbb9f405080482cf9b73cd3ed38b2eb

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: Applied in f8b1ca288cbb9f405080482cf9b73cd3ed38b2eb

**Review by maarquitos14 — APPROVED** [review]
Just a few nits. Otherwise LGTM.

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/more_complex_test_cases.cpp:145]: Not yet, I plan to craft an isolated reproducer and investigate. We likely have a bug in our intel/llvm implementation that in cases like this we don't properly gather all dependencies when creating host accessor

**Review by keryell — COMMENTED** [review]
That looks nice.
Still have this `wait();  // why is it required?` question.

**Comment by AlexeySachkov** [comment]
> Still have this wait();  // why is it required? question.

I will update the PR with a better wording around that comment once I have this issue with dependency tracking reported. I don't think that it should block this PR, because this bug is very likely unrelated to the composite device extension, but the wording we have should definitely be improved before merging. That's still in my TODO-list

**Comment by keryell** [comment]
> > Still have this wait();  // why is it required? question.
> 
> I will update the PR with a better wording around that comment once I have this issue with dependency tracking reported. I don't think that it should block this PR, because this bug is very likely unrelated to the composite device extension, but the wording we have should definitely be improved before merging. That's still in my TODO-list

Otherwise you remove the `wait()` and merge the PR, as a compelling argument to have DPC++ implementation improved before passing next conformance test?

**Comment by AlexeySachkov** [comment]
> > > Still have this wait();  // why is it required? question.
> > 
> > 
> > I will update the PR with a better wording around that comment once I have this issue with dependency tracking reported. I don't think that it should block this PR, because this bug is very likely unrelated to the composite device extension, but the wording we have should definitely be improved before merging. That's still in my TODO-list
> 
> Otherwise you remove the `wait()` and merge the PR, as a compelling argument to have DPC++ implementation improved before passing next conformance test?

Yeah, that's  a better approach. My local testing shows that it is most likely directly related to the extension and not a general issue with dependencies and in that case it is especially better to highlight it.

Removed those `.wait()` calls in 9dcded43d54ced3d2ad1b39121237df01b4cbc1f and resolved merge conflicts, this should be ready to merge once CI passes.

**Review by keryell — APPROVED** [review]
Thanks!

**Pull request merged**
